src/cargo/core/registry.rs

```diff
@@ -312,8 +312,8 @@ impl<'gctx> PackageRegistry<'gctx> {
 
                 if dep.features().len() != 0 || !dep.uses_default_features() {
                     self.source_config.gctx().shell().warn(format!(
-                        "patch for `{}` uses the features mechanism. \
-                        default-features and features will not take effect because the patch dependency does not support this mechanism",
+                        "patch for `{}` attempts to declare features. \
+                        The `features` and `default-features` keys need to appear in a `dependencies` entry, not the `patch` entry.",
```

Comment:
I think a key piece of information is still missing: that the features will be ignored (if I understand correctly).

What do you think of:
```suggestion
                        "ignoring features in patch for `{}`. \
                        Patch-specific `features` and `default-features` are unsupported; instead set these in the `dependencies` entry.",
```

Reply:
imo I think I'd also do this as:
```
shell().warn("ignoring features in patch for `{}`");
shell().note("patch-specific `features` and `default-features` are unsupported; instead set these in the `dependencies` entry");
```

Reply:
@epage I'd like to put the recommendation first, if possible. What about:

```suggestion
                        "patch for `{}` attempts to declare features. \
                        The `features` and `default-features` keys need to appear in a `dependencies` entry; they will be ignored in a `patch` entry.",
```

Reply:
That feels odd to me as this is a warning and should lead with what the warning is about.  I see that being like rustc saying 
```
error: consider introducing a named lifetime parameter

note: missing lifetime specifier
```

I do want to change my suggestion though after thinking more about why these are ignored:
```rust
shell().warn("unused fields in patch for `{}`: {}");
shell().note("configure {} in the `dependencies` entry");
```

Reply:
The warning that this leads with is ```patch for `{}` attempts to declare features```.

But :+1: for the warn+note version you just gave:

```rust
shell().warn("unused fields in patch for `{}`: {}");
shell().note("configure {} in the `dependencies` entry");
```